### PR TITLE
Fix for incorrect sequence of prompts being sent over for inference

### PR DESCRIPTION
**Pull request by ethantang-db**
This fixes the issue where the sequence of prompts we are sending over to vLLM is incorrect and is actually causing a scattering pattern through out its sequence wrt to the unique prompts generations where what we want is that the sequence should have the generations be grouped together for each unique prompt. This issue down the line will cause cases where certain set ups would cause the model to not train as well as this should fix the calculation for its rewards and advantage calculation.

Runs with the fix:

Single node run: https://dbc-559ffd80-2bfc.cloud.databricks.com/ml/experiments/3786561063037191/runs/6921b098214b4b0c88397c5ce0b170a2?o=7395834863327820
2 nodes run: https://dbc-559ffd80-2bfc.cloud.databricks.com/ml/experiments/3786561063037191/runs/a9a74b1a4b2a4561941513d37c305550?o=7395834863327820

### CODE REVIEW

**Review by abaheti95 — APPROVED** [review]
Awesome. This is a fine fix for now while generation per prompt <= rollouts per train actor. But, we would have to do a long term fix once we move the Rewards in the RolloutActor.

File: compose_rl/algorithms/online/callback_utils.py [hunk 1]
```diff
@@ -15,42 +15,45 @@ def preprocess_batches(batches: list, generations_per_prompt: int, pad_token_idx
...
+                        padding_key = False
+
+                    # Compute the required padding and concatenate with the batch tensor
+                    pad = torch.ones(
+                        max_len - seq_len,
+                        dtype=item.dtype,
+                    ) * padding_key  # type: ignore
+                    curr_values.append(torch.cat([pad, item], dim=-1))
 
         # For tensor fields, use torch.cat to combine the values; for string fields, just use the list
```
[the remark this answers is not in the document]
Comment (ethantang-db) [reply compose_rl/algorithms/online/callback_utils.py:54]: This should still be cat, because we are concatenating the padding with the prompt itself to ensure the length will match (ie no jagged arrays)

**Review by rithwik-db — APPROVED** [review]
Left one comment, but LGTM, thanks for catching this!

File: compose_rl/algorithms/online/callback_utils.py [hunk 2]
```diff
@@ -15,42 +15,45 @@ def preprocess_batches(batches: list, generations_per_prompt: int, pad_token_idx
...
-
-                # Compute the required padding and concatenate with the batch tensor
-                pad = torch.ones(
-                    (bs, max_len - seq_len),
-                    dtype=batch[key].dtype,
-                ) * padding_key  # type: ignore
-                curr_values.append(torch.cat([pad, batch[key]], dim=-1))
+            # inside the batch, it's a dictionary of tensors that have the batch dimension there,
+            # so we need to iterate through each element to explode it.
+            for item in batch[key]:
```
Comment (bowenyang008) [inline compose_rl/algorithms/online/callback_utils.py:20]: we need an assertion in the dataloader actor that global_batch_size * num_update_per_iter // num_gen_per_prompt // num_train_gpus is fully divisible

Comment (ethantang-db) [reply compose_rl/algorithms/online/callback_utils.py:20]: I put that in the main function because we don't have access to num_train_gpus in the dataloader actor

File: test_single_controller_ppo.py [hunk 3]
```diff
@@ -768,6 +773,9 @@ def _run_single_controller_ppo(
                 config=config,
             )
 
+            num_prompts_per_iteration = get_and_validate_num_prompts_per_iteration(config)
+            assert num_prompts_per_iteration % num_train_actors == 0, "Number of prompts per iteration must be divisible by number of train actors to ensure accurate advantage calculations."
```
[the remark this answers is not in the document]
Comment (ethantang-db) [reply test_single_controller_ppo.py:777]: Will sync up on more details about this later
